### DISCUSSION TITLE
[OP Stack Deployment] How to build Devnet Instead L1 Sepolia

### DISCUSSION THREAD

**Discussion by avenbreaks** [ethereum-optimism/developers#20] [Deploying the OP Stack] [answered]
### Issue Description

cannot to deploy devnet using make on root directory repositories. 

Logs : 

```
make[1]: Entering directory '/home/ubuntu/optimism'
./ops/scripts/geth-version-checker.sh && \
        (echo "Geth versions match, not installing geth..."; true) || \
                (echo "Versions do not match, installing geth!"; \
                        go install -v github.com/ethereum/go-ethereum/cmd/geth@v1.13.4; \
                        echo "Installed geth!"; true)
./ops/scripts/geth-version-checker.sh: line 4: geth: command not found
Geth version does not match!
geth version: v
.gethrc version: v1.13.4-stable
Versions do not match, installing geth!
Installed geth!
make[1]: Leaving directory '/home/ubuntu/optimism'
./ops/scripts/newer-file.sh .devnet/allocs-l1.json ./packages/contracts-bedrock \
        || make devnet-allocs
make[1]: Entering directory '/home/ubuntu/optimism'
make[2]: Entering directory '/home/ubuntu/optimism'
./ops/scripts/geth-version-checker.sh && \
        (echo "Geth versions match, not installing geth..."; true) || \
                (echo "Versions do not match, installing geth!"; \
                        go install -v github.com/ethereum/go-ethereum/cmd/geth@v1.13.4; \
                        echo "Installed geth!"; true)
./ops/scripts/geth-version-checker.sh: line 4: geth: command not found
Geth version does not match!
geth version: v
.gethrc version: v1.13.4-stable
Versions do not match, installing geth!
Installed geth!
make[2]: Leaving directory '/home/ubuntu/optimism'
PYTHONPATH=./bedrock-devnet python3 ./bedrock-devnet/main.py --monorepo-dir=. --allocs
Traceback (most recent call last):
  File "./bedrock-devnet/main.py", line 1, in <module>
    import devnet
  File "/home/ubuntu/optimism/bedrock-devnet/devnet/__init__.py", line 23, in <module>
    parser.add_argument('--allocs', help='Only create the allocs and exit', type=bool, action=argparse.BooleanOptionalAction)
AttributeError: module 'argparse' has no attribute 'BooleanOptionalAction'
make[1]: *** [Makefile:140: devnet-allocs] Error 1
make[1]: Leaving directory '/home/ubuntu/optimism'
make: *** [Makefile:115: devnet-up] Error 2
```

**Comment by opfocus** [comment]
First, a devnet is a complete local network that includes both L1 and L2. Each time you rebuild it, it starts from scratch. Please follow the guidelines below.     https://github.com/ethereum-optimism/optimism/blob/67a9b3217589cec3f38742ac1f4b83ed27f33cba/specs/meta/devnet.md#bedrock-local-devnet-setup

**Reply by avenbreaks** [reply]
thankyou but i cannot to start devnet-up using make i got the error on python script 

```
make[2]: Leaving directory '/home/ubuntu/optimism'
PYTHONPATH=./bedrock-devnet python3 ./bedrock-devnet/main.py --monorepo-dir=. --allocs
Traceback (most recent call last):
  File "./bedrock-devnet/main.py", line 1, in <module>
    import devnet
  File "/home/ubuntu/optimism/bedrock-devnet/devnet/__init__.py", line 23, in <module>
    parser.add_argument('--allocs', help='Only create the allocs and exit', type=bool, action=argparse.BooleanOptionalAction)
AttributeError: module 'argparse' has no attribute 'BooleanOptionalAction'
make[1]: *** [Makefile:140: devnet-allocs] Error 1
make[1]: Leaving directory '/home/ubuntu/optimism'
make: *** [Makefile:115: devnet-up] Error 2
```

**Reply by opfocus** [reply]
make sure that you have compiled the contracts by cding into packages/contracts-bedrock and running pnpm i followed by pnpm build

**Reply by Chomtana** [reply]
Have you done

pnpm install
pnpm build

If not working you may try to revert back to v1.1.4 (latest known stable) by

git checkout v1.1.4

**Reply by avenbreaks** [reply]
thankyou will try again

**Reply by avenbreaks** [reply]
same error 

```
make[1]: Entering directory '/home/ubuntu/optimism'
PYTHONPATH=./bedrock-devnet python3 ./bedrock-devnet/main.py --monorepo-dir=. --allocs
Traceback (most recent call last):
  File "./bedrock-devnet/main.py", line 1, in <module>
    import devnet
  File "/home/ubuntu/optimism/bedrock-devnet/devnet/__init__.py", line 20, in <module>
    parser.add_argument('--allocs', help='Only create the allocs and exit', type=bool, action=argparse.BooleanOptionalAction)
AttributeError: module 'argparse' has no attribute 'BooleanOptionalAction'
make[1]: *** [Makefile:112: devnet-allocs] Error 1
make[1]: Leaving directory '/home/ubuntu/optimism'
make: *** [Makefile:90: devnet-up] Error 2
```

Python 3.8.10

**Reply by opfocus** [reply]
Can you please try my suggestion, which is to comment out the line in the Makefile that reads ./ops/scripts/newer-file.sh .devnet/allocs-l1.json ./packages/contracts-bedrock || make devnet-allocs? Typically, when we follow the steps, I believe this statement will not be executed.
Previously, there was an issue with the Python version, but it seems to have been resolved in the latest commit. If you suspect a Python version issue, please try the following solutions:

Python 3.8 was the default in Ubuntu 20.04 LTS, but we need Python 3.9 here.

Steps to reproduce:

Steps to fix:

bash
Copy code
# Install Python 3.9
sudo apt update
sudo apt install software-properties-common
sudo add-apt-repository ppa:deadsnakes/ppa
sudo apt update
sudo apt install python3.9
python3.9 -V

# Replace 'python3' with 'python3.9' in the makefile

or

# Install Python 3.9
sudo apt install python3.9

# Set Python 3.9 as the default
sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.8 1
sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.9 2
sudo update-alternatives --config python3
```"

**Reply by opfocus** [reply]
My most direct suggestion is to re-git clone the optimism repository and then follow the steps   https://github.com/ethereum-optimism/optimism/blob/67a9b3217589cec3f38742ac1f4b83ed27f33cba/specs/meta/devnet.md#bedrock-local-devnet-setup

**Reply by opfocus** [reply]
Okay, let's try it together

**Reply by Chomtana** [reply, marked as answer]
https://stackoverflow.com/questions/70242395/run-capital-gains-command-and-get-erro-attributeerror-module-argparse-has-no

Need python 3.9+ to work

Confirmed python version issue. I recommended 3.10+ or upgrade ubuntu to 22.04

**Comment by opfocus** [comment]
# Setting Up and Running a Devnet Tutorial

In this tutorial, we'll guide you through setting up and running a devnet using the latest commit for Ethereum Optimism(commit a37f81dc4b29eb6f4570ee62d5a6412230ff35c1). A devnet is a local development network for Ethereum Layer 1 (L1) and Layer 2 (L2) testing.

## Prerequisites

- Ubuntu-based Linux system.
- A working installation of [Docker Engine](https://docs.docker.com/engine/install/ubuntu).
- Basic knowledge of using a terminal.

### Step 1: Clone the Repository

Open your terminal and run the following command to clone the Ethereum Optimism repository:

```bash
cd ~
git clone https://github.com/ethereum-optimism/optimism.git
```
## Step 2: Navigate to the Contracts Bedrock Directory

Change your working directory to the Contracts Bedrock directory within the cloned repository:

```bash
cd optimism/packages/contracts-bedrock
```
### Step 3: Install Dependencies

Use [pnpm](https://pnpm.io/) to install the project dependencies:

```bash
pnpm install
```
### Step 4: Build the Project

Compile the project by running:

```bash
pnpm build
```
### Step 5: Start the Devnet

Now, it's time to launch the devnet. Use the following command:

```bash
make devnet-up
```
### Step 6: Add 'geth' to the PATH

Edit your `.bashrc` file to add the 'geth' executable's path. Use a text editor or run:

```bash
echo 'export PATH="$PATH:/root/go/bin"' >> ~/.bashrc
```
```bash
source ~/.bashrc
```

### Step 7: Ensure Docker is Installed

If you face issues related to missing Docker files, ensure that you have Docker Engine installed on your Ubuntu system. Follow the official Docker installation guide for Ubuntu [here](https://docs.docker.com/engine/install/ubuntu).

### Step 8: Run Devnet Again

After completing the above steps, you should now be able to successfully run the devnet using:

```bash
make devnet-up
```

### Additional Information:

If your system is pre-installed with Python 3.8 and you encounter issues, you may need to install Python 3.9 and set it as the default. Follow these steps:

Install Python 3.9:
```bash
sudo apt install python3.9
```
Set Python 3.9 as the default:
```bash
sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.8 1
sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.9 2
sudo update-alternatives --config python3
```

**Comment by avenbreaks** [comment]
solved thankyou
